File: ci/jimm_ci/github_app.py
"""Minimal GitHub App client.

Handles JWT minting, installation token caching, the slice of the Checks
API we need, the Compare API for path-filter input, and small helpers for
listing PRs / commits / check-runs and applying labels. Uses httpx for
async I/O and PyJWT for RS256 signing. No other GitHub libraries.
"""
from __future__ import annotations

import logging
import time
from dataclasses import dataclass
from typing import Any

import httpx
import jwt

LOG = logging.getLogger(__name__)
API = "https://api.github.com"


@dataclass
class CheckRun:
    id: int
    name: str
    head_sha: str
    html_url: str


class GitHubApp:
    def __init__(
        self,
        app_id: int,
        installation_id: int,
        private_key: str,
        *,
        client: httpx.AsyncClient | None = None,
    ) -> None:
        self.app_id = app_id
        self.installation_id = installation_id
        self.private_key = private_key
        self._client = client or httpx.AsyncClient(timeout=30.0)
        self._token: str | None = None
        self._token_exp: float = 0.0

    async def aclose(self) -> None:
        await self._client.aclose()

    def _mint_jwt(self) -> str:
        now = int(time.time())
        return jwt.encode(
            {"iat": now - 60, "exp": now + 540, "iss": str(self.app_id)},
            self.private_key,
            algorithm="RS256",
        )

    async def installation_token(self) -> str:
        now = time.time()
        if self._token and now < self._token_exp - 300:
            return self._token
        app_jwt = self._mint_jwt()
        r = await self._client.post(
            f"{API}/app/installations/{self.installation_id}/access_tokens",
            headers={
                "Authorization": f"Bearer {app_jwt}",
                "Accept": "application/vnd.github+json",
                "X-GitHub-Api-Version": "2022-11-28",
            },
        )
        r.raise_for_status()
        data = r.json()
        self._token = data["token"]
        self._token_exp = now + 3300
        return self._token

    async def _request(self, method: str, path: str, **kwargs: Any) -> httpx.Response:
        token = await self.installation_token()
        headers = kwargs.pop("headers", {})
        headers.setdefault("Authorization", f"Bearer {token}")
        headers.setdefault("Accept", "application/vnd.github+json")
        headers.setdefault("X-GitHub-Api-Version", "2022-11-28")
        r = await self._client.request(method, f"{API}{path}", headers=headers, **kwargs)
        if r.status_code == 401:
            self._token = None
            self._token_exp = 0
            token = await self.installation_token()
            headers["Authorization"] = f"Bearer {token}"
            r = await self._client.request(method, f"{API}{path}", headers=headers, **kwargs)
        r.raise_for_status()
        return r

    async def _paginated(self, path: str, *, per_page: int = 100,
                         max_pages: int = 20) -> list[dict[str, Any]]:
        sep = "&" if "?" in path else "?"
        url: str | None = f"{path}{sep}per_page={per_page}"
        out: list[dict[str, Any]] = []
        pages = 0
        while url and pages < max_pages:
            if url.startswith("http"):
                token = await self.installation_token()
                r = await self._client.get(
                    url,
                    headers={
                        "Authorization": f"Bearer {token}",
                        "Accept": "application/vnd.github+json",
                        "X-GitHub-Api-Version": "2022-11-28",
                    },
                )
                r.raise_for_status()
            else:
                r = await self._request("GET", url)
            page = r.json()
            if not isinstance(page, list):
                break
            out.extend(page)
            next_link = r.links.get("next") if hasattr(r, "links") else None
            url = next_link["url"] if next_link else None
            pages += 1
        return out

    async def create_check_run(
        self,
        repo: str,
        head_sha: str,
        name: str,
        *,
        status: str = "in_progress",
        conclusion: str | None = None,
        details_url: str | None = None,
        output: dict[str, Any] | None = None,
    ) -> CheckRun:
        body: dict[str, Any] = {"name": name, "head_sha": head_sha, "status": status}
        if conclusion is not None:
            body["conclusion"] = conclusion
        if details_url is not None:
            body["details_url"] = details_url
        if output is not None:
            body["output"] = output
        r = await self._request("POST", f"/repos/{repo}/check-runs", json=body)
        data = r.json()
        return CheckRun(
            id=data["id"], name=data["name"], head_sha=data["head_sha"],
            html_url=data["html_url"],
        )

    async def complete_check_run(
        self,
        repo: str,
        check_run_id: int,
        *,
        conclusion: str,
        output: dict[str, Any] | None = None,
    ) -> None:
        body: dict[str, Any] = {"status": "completed", "conclusion": conclusion}
        if output is not None:
            body["output"] = output
        await self._request("PATCH", f"/repos/{repo}/check-runs/{check_run_id}", json=body)

    async def compare(self, repo: str, base: str, head: str) -> list[str]:
        """Return the list of changed file paths between base and head."""
        r = await self._request("GET", f"/repos/{repo}/compare/{base}...{head}")
        data = r.json()
        return [f["filename"] for f in data.get("files", [])]

    async def get_default_branch_head(self, repo: str, branch: str = "master") -> str:
        r = await self._request("GET", f"/repos/{repo}/branches/{branch}")
        return r.json()["commit"]["sha"]

    async def list_open_pulls(self, repo: str) -> list[dict[str, Any]]:
        """List open PRs with head/base, labels, and fork status."""
        return await self._paginated(f"/repos/{repo}/pulls?state=open")

    async def list_commits(self, repo: str, *, sha: str, since: str) -> list[dict[str, Any]]:
        """List commits on `sha` (branch or commit) committed at or after `since`.

        `since` must be an ISO-8601 timestamp (e.g. `2026-04-20T00:00:00Z`).
        """
        return await self._paginated(
            f"/repos/{repo}/commits?sha={sha}&since={since}",
        )

    async def list_check_runs(self, repo: str, ref: str) -> list[dict[str, Any]]:
        """Return all check runs for `ref`. Paginated under `check_runs`."""
        out: list[dict[str, Any]] = []
        per_page = 100
        page = 1
        while True:
            r = await self._request(
                "GET",
                f"/repos/{repo}/commits/{ref}/check-runs"
                f"?per_page={per_page}&page={page}",
            )
            data = r.json()
            runs = data.get("check_runs", []) if isinstance(data, dict) else []
            out.extend(runs)
            if len(runs) < per_page:
                return out
            page += 1
            if page > 20:
                return out

File: ci/JimmCI/src/GitHubApp.jl
module GitHubAppMod

using Dates
using HTTP
using JSON3
using JSONWebTokens

export GitHubApp,
    create_check_run,
    complete_check_run,
    compare,
    get_default_branch_head,
    list_open_pulls,
    list_commits,
    list_check_runs,
    CheckRun

const API = "https://api.github.com"

struct CheckRun
    id::Int
    name::String
    head_sha::String
    html_url::String
end

mutable struct GitHubApp
    app_id::Int
    installation_id::Int
    private_key::String
    token::Union{String,Nothing}
    token_exp::Float64
    lock::ReentrantLock
end

GitHubApp(app_id::Integer, installation_id::Integer, private_key::AbstractString) =
    GitHubApp(
        Int(app_id),
        Int(installation_id),
        String(private_key),
        nothing,
        0.0,
        ReentrantLock(),
    )

# ── Auth ──────────────────────────────────────────────────────────────

function _mint_jwt(gh::GitHubApp)
    now_s = floor(Int, time())
    claims = Dict("iat" => now_s - 60, "exp" => now_s + 540, "iss" => string(gh.app_id))
    enc = JSONWebTokens.RS256(gh.private_key)
    return JSONWebTokens.encode(enc, claims)
end

function _refresh_token!(gh::GitHubApp)
    app_jwt = _mint_jwt(gh)
    r = HTTP.request(
        "POST",
        string(API, "/app/installations/", gh.installation_id, "/access_tokens");
        headers = [
            "Authorization" => "Bearer $app_jwt",
            "Accept" => "application/vnd.github+json",
            "X-GitHub-Api-Version" => "2022-11-28",
        ],
        status_exception = true,
    )
    data = JSON3.read(String(r.body))
    gh.token = String(data["token"])
    # GitHub-issued installation tokens are valid for ~1h; keep a 5m safety margin.
    gh.token_exp = time() + 3300
    return gh.token
end

function installation_token(gh::GitHubApp)
    @lock gh.lock begin
        if gh.token !== nothing && time() < gh.token_exp - 300
            return gh.token
        end
        return _refresh_token!(gh)
    end
end

# ── Core request helper ───────────────────────────────────────────────

function _auth_headers(gh::GitHubApp; extra = ())
    token = installation_token(gh)
    base = [
        "Authorization" => "Bearer $token",
        "Accept" => "application/vnd.github+json",
        "X-GitHub-Api-Version" => "2022-11-28",
    ]
    return vcat(base, collect(extra))
end

function _request(
    gh::GitHubApp,
    method::AbstractString,
    path::AbstractString;
    body = nothing,
    query = nothing,
)
    url = startswith(path, "http") ? path : string(API, path)
    headers = _auth_headers(gh)
    opts = Dict{Symbol,Any}(:headers => headers, :status_exception => false)
    if body !== nothing
        opts[:body] = JSON3.write(body)
        push!(headers, "Content-Type" => "application/json")
    end
    if query !== nothing
        opts[:query] = query
    end
    r = HTTP.request(method, url; opts...)
    if r.status == 401
        # Drop cached token and retry once.
        @lock gh.lock begin
            gh.token = nothing
            gh.token_exp = 0.0
        end
        opts[:headers] = _auth_headers(gh)
        r = HTTP.request(method, url; opts...)
    end
    if r.status >= 400
        msg = String(r.body)
        if length(msg) > 500
            msg = msg[1:500] * "…"
        end
        error("GitHub $method $path failed: HTTP $(r.status) $msg")
    end
    return r
end

# ── Pagination ────────────────────────────────────────────────────────

# Extract the `next` link from a GitHub Link header. The header looks
# like:  <https://api.github.com/foo?page=2>; rel="next", <…>; rel="last"
function _next_link(r::HTTP.Response)
    link = HTTP.header(r, "Link")
    isempty(link) && return nothing
    for part in split(link, ',')
        m = match(r"^\s*<([^>]+)>\s*;\s*rel=\"([^\"]+)\"\s*$", part)
        m === nothing && continue
        if m.captures[2] == "next"
            return String(m.captures[1])
        end
    end
    return nothing
end

function _paginated(
    gh::GitHubApp,
    path::AbstractString;
    per_page::Int = 100,
    max_pages::Int = 20,
)
    sep = occursin('?', path) ? '&' : '?'
    url = string(path, sep, "per_page=", per_page)
    out = Any[]
    pages = 0
    while url !== nothing && pages < max_pages
        r = _request(gh, "GET", url)
        page = JSON3.read(String(r.body))
        if !(page isa AbstractVector)
            break
        end
        append!(out, page)
        url = _next_link(r)
        pages += 1
    end
    return out
end

# ── Checks API ────────────────────────────────────────────────────────

function create_check_run(
    gh::GitHubApp,
    repo::AbstractString,
    head_sha::AbstractString,
    name::AbstractString;
    status::AbstractString = "in_progress",
    conclusion::Union{Nothing,AbstractString} = nothing,
    details_url::Union{Nothing,AbstractString} = nothing,
    output = nothing,
)
    body = Dict{String,Any}("name" => name, "head_sha" => head_sha, "status" => status)
    conclusion === nothing || (body["conclusion"] = conclusion)
    details_url === nothing || (body["details_url"] = details_url)
    output === nothing || (body["output"] = output)
    r = _request(gh, "POST", "/repos/$repo/check-runs"; body = body)
    data = JSON3.read(String(r.body))
    return CheckRun(
        Int(data["id"]),
        String(data["name"]),
        String(data["head_sha"]),
        String(data["html_url"]),
    )
end

function complete_check_run(
    gh::GitHubApp,
    repo::AbstractString,
    check_run_id::Integer;
    conclusion::AbstractString,
    output = nothing,
)
    body = Dict{String,Any}("status" => "completed", "conclusion" => conclusion)
    output === nothing || (body["output"] = output)
    _request(gh, "PATCH", "/repos/$repo/check-runs/$check_run_id"; body = body)
    return nothing
end

# ── Compare / Pulls / Commits / Check Runs listing ────────────────────

"""Return the list of changed file paths between `base` and `head`."""
function compare(
    gh::GitHubApp,
    repo::AbstractString,
    base::AbstractString,
    head::AbstractString,
)
    r = _request(gh, "GET", "/repos/$repo/compare/$base...$head")
    data = JSON3.read(String(r.body))
    files = get(data, "files", nothing)
    files === nothing && return String[]
    return [String(f["filename"]) for f in files]
end

function get_default_branch_head(
    gh::GitHubApp,
    repo::AbstractString,
    branch::AbstractString = "master",
)
    r = _request(gh, "GET", "/repos/$repo/branches/$branch")
    data = JSON3.read(String(r.body))
    return String(data["commit"]["sha"])
end

function list_open_pulls(gh::GitHubApp, repo::AbstractString)
    return _paginated(gh, "/repos/$repo/pulls?state=open")
end

"""
List commits on `sha` (branch or commit) committed at or after `since`,
where `since` is an ISO-8601 timestamp (e.g. `2026-04-20T00:00:00Z`).
"""
function list_commits(
    gh::GitHubApp,
    repo::AbstractString;
    sha::AbstractString,
    since::AbstractString,
)
    return _paginated(gh, "/repos/$repo/commits?sha=$sha&since=$since")
end

"""Return all check runs for `ref`. Result paginated under `check_runs`."""
function list_check_runs(gh::GitHubApp, repo::AbstractString, ref::AbstractString)
    out = Any[]
    per_page = 100
    page = 1
    while true
        r = _request(
            gh,
            "GET",
            "/repos/$repo/commits/$ref/check-runs?per_page=$per_page&page=$page",
        )
        data = JSON3.read(String(r.body))
        runs = get(data, "check_runs", nothing)
        runs === nothing && return out
        append!(out, runs)
        length(runs) < per_page && return out
        page += 1
        page > 20 && return out
    end
end

end # module
